### ISSUE TITLE
Listing from the book is incompatible with the current stable release of gtk4-rs.

### ISSUE THREAD

**Issue by AbrarSL** [gtk-rs/gtk4-rs#1229]
In this [listing](https://github.com/gtk-rs/gtk4-rs/blob/master/book/listings/list_widgets/4/main.rs) line 39, in the signal handler for the `SignalListItemFactory`,
```rs
// ANCHOR: factory_setup
    factory.connect_setup(move |_, list_item| {
        // Create label
        let label = Label::new(None);
        list_item.set_child(Some(&label));

        // Bind `list_item->item->number` to `label->label`
        list_item
            .property_expression("item")
            .chain_property::<IntegerObject>("number")
            .bind(&label, "label", Widget::NONE);
    });
    // ANCHOR_END: factory_setup
```
No longer works because of an API change for the callbacks for [SignalListItemFactory](https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4/struct.SignalListItemFactory.html#method.connect_setup), the item passed to the callback is a `glib::object::Object` that needs to be downcast before `ListItem` methods such as `set_child()` can be used.

```rs
 let list_item = list_item
            .downcast_ref::<ListItem>()
            .expect("Needs to be ListItem");
```

**Comment by Hofer-Julian** [comment]
Thanks for opening the issue @AbrarSL.
Are you interested in opening a PR to fix this?

**Comment by Hofer-Julian** [comment]
I see the problem now.
That is the API change that only comes into action if the crate feature `v4_8` is activated.
I tend towards assuming that no crate feature is activated and adding an info box where [SignalListItemFactory](https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4/struct.SignalListItemFactory.html#method.connect_setup) is used.
What do you say @bilelmoussaoui?

**Comment by AbrarSL** [comment]
> Thanks for opening the issue @AbrarSL. Are you interested in opening a PR to fix this?

I wouldn't mind doing that if I just had to change the code snippet but this looks like it needs changes to the text as well? I don't think I should be doing that :sweat_smile:

**Comment by GuillaumeGomez** [comment]
Mail test for @sdroege.

**Comment by Hofer-Julian** [comment]
@AbrarSL Yeah, you can leave it. I thought it's just a breakage but seems it's more complicated than that.

**Comment by bilelmoussaoui** [comment]
> I see the problem now. That is the API change that only comes into action if the crate feature `v4_8` is activated. I tend towards assuming that no crate feature is activated and adding an info box where [SignalListItemFactory](https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4/struct.SignalListItemFactory.html#method.connect_setup) is used. What do you say @bilelmoussaoui?


I would make use of the v4_8 stuff by default and add an information about the older version instead.

**Comment by Hofer-Julian** [comment]
> > I see the problem now. That is the API change that only comes into action if the crate feature `v4_8` is activated. I tend towards assuming that no crate feature is activated and adding an info box where [SignalListItemFactory](https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4/struct.SignalListItemFactory.html#method.connect_setup) is used. What do you say @bilelmoussaoui?
> 
> I would make use of the v4_8 stuff by default and add an information about the older version instead.

I avoided that until now, because I haven't used any functions behind a feature gate in the book.
However, I assume that most people use a `v4_*` in practice anyway, so I guess we can add it too.

**Comment by sdroege** [comment]
API changes based on feature flags is a very bad idea, don't do that but instead add new API that is enabled by the feature flag.

**Comment by bilelmoussaoui** [comment]
well we can't do much here as things would break at runtime as the signals would pass you a glib::Object instead of a gtk::ListItem if the gtk version is >= 4.8

**Comment by sdroege** [comment]
But that actual object in the value would still be a `gtk::ListItem`, so that's going to work fine.

**Comment by sdroege** [comment]
Oh I see, but not if you use the old version of the signal API. Well that sucks and arguably shouldn't have happened in C (it's going to be a problem for other statically typed languages too).
